### PR TITLE
Made functions members of respective classes rather than passing references to object of class

### PR DESCRIPTION
**Pull request by adisidev**
This pull request relies on #117, as I changed the kernel for this pull request too. Thus, it may be a good idea to check that out first. This addresses multiple issues, which I shall list out shortly.

### CODE REVIEW

**Review by mgastner — CHANGES_REQUESTED** [review]
This pull request looks very nice. I only have a few comments and changes for requests. Please see my comments in the code.

File: src/inset_state/albers_projection.cpp [hunk 1]
```diff
@@ -128,41 +126,17 @@ void transform_to_albers_projection(InsetState *inset_state)
...
-          coords_hole = projected_albers_coordinates(coords_hole,
-                                                     lambda_0,
-                                                     phi_0,
-                                                     phi_1,
-                                                     phi_2);
-        }
-      }
-    }
-  }
+  // Specialise point_after_albers_projection with lx_ and ly_
```
[the remark this answers is not in the document]
Comment (adisidev) [reply src/inset_state/albers_projection.cpp:129]: I've changed the comment, and added the word `curry` as in https://en.wikipedia.org/wiki/Currying

More can be found here, upon searching the keyword "specialize": https://dev.realworldocaml.org/variables-and-functions.html

File: src/inset_state/project.cpp [hunk 2]
```diff
@@ -1,105 +1,74 @@
...
-      const Polygon_with_holes new_pwh(new_ext_ring,
-                                       hole_v.begin(),
-                                       hole_v.end());
-      new_gd.push_back(new_pwh);
-    }
-    new_geo_divs.push_back(new_gd);
-  }
-  inset_state->set_geo_divs(new_geo_divs);
+
+  // Specialise interpolate_point_bilinearly
```
[the remark this answers is not in the document]
Comment (adisidev) [reply src/inset_state/project.cpp]: Same as the comment about the word specialise earlier.

File: src/inset_state/project.cpp [hunk 3]
```diff
@@ -409,67 +364,59 @@ Point affine_trans(const std::array<Point, 3> tri,
...
 
-          // Update hole coordinates
-          const auto new_hole_pt =
-            projected_point_with_triangulation((*h)[i], inset_state);
-          new_hole.push_back(new_hole_pt);
+        // Iterate over hole's coordinates
+        for (auto &coords_hole : *h) {
+
+          // Assign hole's coordinates to transformed coordinates
+          coords_hole = projected_point_with_triangulation(coords_hole);
```
[the remark this answers is not in the document]
Comment (adisidev) [reply src/inset_state/project.cpp:405]: Indeed, it very much could be; I've changed it as suggested, thanks for point it out.

[the remark this answers is not in the document]
Comment (adisidev) [reply src/inset_state/project.cpp:405]: Alas, I realise why I did not do it initially. After some more research, I was able to implement it as requested and reduce the unnecessary code-duplication.

File: src/inset_state/rescale_map.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  double new_xmin =
+    0.5 * ((1.0-padding)*bb.xmax() + (1.0+padding)*bb.xmin());
+  double new_xmax =
+    0.5 * ((1.0+padding)*bb.xmax() + (1.0-padding)*bb.xmin());
+  double new_ymin =
+    0.5 * ((1.0-padding)*bb.ymax() + (1.0+padding)*bb.ymin());
+  double new_ymax =
+    0.5 * ((1.0+padding)*bb.ymax() + (1.0-padding)*bb.ymin());
+
+  // Ensure that the grid dimensions lx and ly are integer powers of 2
```
[the remark this answers is not in the document]
Comment (adisidev) [reply src/inset_state/rescale_map.cpp:31]: I'm unsure whether we should remove them or not. I have shifted `unsigned int lx, ly;` closer to the comment. Indeed, the next few lines do deal with setting appropriate `lx` and `ly`.  However, I do understand your concern too. So I shall leave it up to you to decide later.

[the remark this answers is not in the document]
Comment (adisidev) [reply src/inset_state/rescale_map.cpp:31]: Instead, we could also round up to the nearest power of 2. However, doing this implicitly may confuse users why their sizes were changed.

**Comment by adisidev** [comment]
Lots of functions are now unused. We should decide whether to keep them or not. I will create a commit removing them, and we can revert the commit if we feel it is best to retain the unused functions.

**Review by mgastner — APPROVED** [review]
Thank you for sorting the files. I also like the lambda-function approach to transforming points in `albers_projection.cpp`, `project.cpp` and `smyth_projection.cpp`. We should slowly roll out this technique broadly in similar cases.

**Pull request merged**